mesonbuild/mparser.py

```diff
@@ -72,22 +76,22 @@ def __init__(self, text, line, lineno, colno, start_line, start_lineno, start_co
         self.colno = colno
 
 class Token:
-    def __init__(self, tid, subdir, line_start, lineno, colno, bytespan, value):
-        self.tid = tid
-        self.subdir = subdir
-        self.line_start = line_start
-        self.lineno = lineno
-        self.colno = colno
-        self.bytespan = bytespan
-        self.value = value
+    def __init__(self, tid: str, subdir: str, line_start: int, lineno: int, colno: int, bytespan: Tuple[int, int], value: Union[str, bool, int]) -> None:
+        self.tid = tid                # type: str
+        self.subdir = subdir          # type: str
+        self.line_start = line_start  # type: int
+        self.lineno = lineno          # type: int
+        self.colno = colno            # type: int
+        self.bytespan = bytespan      # type: Tuple[int, int]
+        self.value = value            # type: Union[str, bool, int]
```

Comment:
I think this would be better expressed as a generic, something like:

```python
TokenTypes = typing.TypeVar('TokenTypes', int, str, bool)

class Token(typing.Generic[TokenTypes)

    def __init__(... value: TokenTypes): ...
```

Also, you shouldn't need to type the assignments here, mypy should be able to figure that out automatically.

Reply:
Done